istioctl/cmd/workload.go

```diff
@@ -0,0 +1,402 @@
+	clusterEnv := map[string]string{
+		"ISTIO_CP_AUTH":       "MUTUAL_TLS",
+		"ISTIO_INBOUND_PORTS": portBehavior,
+		"ISTIO_NAMESPACE":     wg.Namespace,
+		"ISTIO_PILOT_PORT":    "15012",
+		"ISTIO_SERVICE":       fmt.Sprintf("%s.%s", wg.Name, wg.Namespace),
+		"ISTIO_SERVICE_CIDR":  "*",
+		"SERVICE_ACCOUNT":     we.ServiceAccount,
+	}
+	return ioutil.WriteFile(filepath.Join(dir, "cluster.env"), []byte(mapToString(clusterEnv)), filePerms)
+}
+
+// Get and store the needed certificate and token. The certificate comes from the CA root cert, and
+// the token is generated by kubectl under the workload group's namespace and service account
+// TODO: Make the following accurate when using the Kubernetes certificate signer
+func createCertsTokens(kubeClient kube.ExtendedClient, wg *clientv1alpha3.WorkloadGroup, dir string) error {
+	rootCert, err := kubeClient.CoreV1().ConfigMaps(wg.Namespace).Get(context.Background(), controller.CACertNamespaceConfigMap, metav1.GetOptions{})
+	// errors if the requested configmap does not exist in the given namespace
+	if err != nil {
+		return fmt.Errorf("configmap %s was not found in namespace %s: %v", controller.CACertNamespaceConfigMap, wg.Namespace, err)
+	}
+	if err = ioutil.WriteFile(filepath.Join(dir, "root-cert"), []byte(rootCert.Data[constants.CACertNamespaceConfigMapDataName]), filePerms); err != nil {
+		return err
+	}
+
+	token := &authenticationv1.TokenRequest{
+		Spec: authenticationv1.TokenRequestSpec{
+			Audiences:         []string{"istio-ca"},
+			ExpirationSeconds: &tokenDuration,
+		},
+	}
+	serviceAccount := wg.Spec.Template.ServiceAccount
+	tokenReq, err := kubeClient.CoreV1().ServiceAccounts(wg.Namespace).CreateToken(context.Background(), serviceAccount, token, metav1.CreateOptions{})
+	// errors if the token could not be created with the given service account in the given namespace
+	if err != nil {
+		return fmt.Errorf("could not create a token under service account %s in namespace %s: %v", serviceAccount, wg.Namespace, err)
+	}
+	if err := ioutil.WriteFile(filepath.Join(dir, "istio-token"), []byte(tokenReq.Status.Token), filePerms); err != nil {
+		return err
+	}
+	return nil
+}
+
+// TODO: Support the proxy.istio.io/config annotation
+func createMeshConfig(kubeClient kube.ExtendedClient, wg *clientv1alpha3.WorkloadGroup, clusterID, revision, dir string) (*meshconfig.MeshConfig, error) {
+	istioCM := "istio"
+	// Case with multiple control planes
+	if revision != "" {
+		istioCM = fmt.Sprintf("%s-%s", istioCM, revision)
+	}
+	istio, err := kubeClient.CoreV1().ConfigMaps(istioNamespace).Get(context.Background(), istioCM, metav1.GetOptions{})
+	// errors if the requested configmap does not exist in the given namespace
+	if err != nil {
+		return nil, fmt.Errorf("configmap %s was not found in namespace %s: %v", istioCM, istioNamespace, err)
+	}
+	meshConfig, err := mesh.ApplyMeshConfigDefaults(istio.Data[configMapKey])
```

Comment:
nit: would be good if we only output the diffs relative to the defaults. I don't mean some complex logic to do this, just not using `ApplyMeshConfigDefaults` I think should do the trick. The rationale is if we have something using the default option we probably want it to be unset so if we change the default they are not unintentionally using the old field